Added language as a filter to explore/repo_search

snematoda opened the pull request:
|      |     |
| --- | --- |
| ![mobile](https://github.com/go-gitea/gitea/assets/122038334/a1a9ea1b-1669-4bc4-a174-ad92d54be77c) | ![landscape](https://github.com/go-gitea/gitea/assets/122038334/2d4fc458-5c45-458a-a38a-7a7f29d6eb75) |

Added a new language filter beside the existing sort for  repos in explore

JakobDev commented on templates/explore/repo_search.tmpl:
```diff
@@ -20,8 +20,21 @@
 	</div>
+
+	<!-- Language -->
+	<div class="ui right dropdown type jump item">
+		<span class="text">Language {{svg "octicon-triangle-down" 14 "dropdown icon"}}</span>
+		<div class="menu">
+			<a class="item" href="{{$.Link}}?q={{$.Keyword}}&sort={{$.SortType}}">Any</a>
+			{{range $lang, $color := $.Languages}}
+			<a class="{{if eq $.Language $lang}}active {{end}}item" href="{{$.Link}}?q={{$.Keyword}}&language={{$lang}}&sort={{$.SortType}}">
+				<span class="gt-df gt-ac gt-mr-3"><i class="color-icon gt-mr-3" style="background-color: {{$color}}"></i>{{$lang}}</span>
+			</a>
+			{{end}}
+		</div>
+	</div>
```
Is this really suitable? The big Problem I was facing, was that this list can get extremely large, so I needed to find another solution. I tried to rewrite the Author search Box from the Issue page for languages, but I don't got it to work.

snematoda replied:
Now that I think about it, this probably isn't the right way to implement this :/

JakobDev replied:
Here's my Go function of collectiong all Languages used in the Instance:
```go
func GetPrimaryRepoLanguageList(ctx context.Context) (LanguageStatList, error) {
	languageList := make(LanguageStatList, 0)

	err := db.GetEngine(ctx).Table("language_stat").Cols("language").Where(builder.Eq{"is_primary": true}).Distinct("language").OrderBy("language").Find(&languageList)

	if err != nil {
		return nil, err
	}

	languageList.LoadAttributes()
	return languageList, nil
}
```
Put it in `models/repo/repo_list.go`.

> I felt that computing the list of languages for all the repos (while insignificant for small instances) might potentially be a bottleneck for larger instances

I have no Way to test this on a very large instance, but considering the language Column is indexed, it should not be that big of a deal. It is still better than showing a List of all Languages where in most cases not more than 10 or so are used.

snematoda replied:
Thank you!

JakobDev commented on routers/web/explore/repo.go:
```diff
@@ -133,6 +135,7 @@ func RenderRepoSearch(ctx *context.Context, opts *RepoSearchOptions) {
 	ctx.Data["Total"] = count
 	ctx.Data["Repos"] = repos
 	ctx.Data["IsRepoIndexerEnabled"] = setting.Indexer.RepoIndexerEnabled
+	ctx.Data["Languages"] = enry.LanguagesColor
```
You should only list the Languages that are used on this Gitea instance. There are a lot of small Instances which has only a few languages, so there is no need to have a full blown list. I can give you my Code if you want.

snematoda replied:
Seems like you already have a better implementation than mine. If you would like to create a new PR, I would be quite pleased to close mine in favor of yours :)

JakobDev replied:
You can keep working on this, if you want.

JakobDev reviewed (commented):
I was already working on this and almost finished, but you were faster. I have left a few comments with the toughs I was facing during developing this.

silverwind commented:
Any screenshot?

snematoda commented:
> Any screenshot?

Sure!
Just updated the PR message

lunny commented on models/repo/repo_list.go:
```diff
@@ -730,3 +730,21 @@ func GetUserRepositories(opts *SearchRepoOptions) (RepositoryList, int64, error)
 }
+
+func GetPrimaryRepoLanguageList(ctx context.Context) (LanguageStatList, error) {
+	languageList := make(LanguageStatList, 0)
```
Since we just need `language`, we can use a string slice here directly.

lunny replied:
This is still an improvement.

JakobDev replied:
The Colour is also needed, not only the name

lunny commented on models/repo/repo_list.go:
```diff
@@ -730,3 +730,21 @@ func GetUserRepositories(opts *SearchRepoOptions) (RepositoryList, int64, error)
 	repos := make(RepositoryList, 0, opts.PageSize)
 	return repos, count, db.SetSessionPagination(sess, opts).Find(&repos)
 }
+
+func GetPrimaryRepoLanguageList(ctx context.Context) (LanguageStatList, error) {
+	languageList := make(LanguageStatList, 0)
+
+	err := db.GetEngine(ctx).
+		Table("language_stat").
+		Cols("language").
+		Where(builder.Eq{"is_primary": true}).
+		Distinct("language").
+		OrderBy("language").
+		Find(&languageList)
+	if err != nil {
+		return nil, err
+	}
+
+	languageList.LoadAttributes()
+	return languageList, nil
+}
```
```suggestion

func GetPrimaryRepoLanguageList(ctx context.Context) ([]string, error) {
	languageList := make([]string, 0)
  err := db.GetEngine(ctx).
		Table("language_stat").
		Cols("language").
		Where(builder.Eq{"is_primary": true}).
		Distinct("language").
		OrderBy("language").
		Find(&languageList)
	return languageList, err
}
```

JakobDev replied:
We also need the Color, which we get with `LoadAttributes()`

snematoda replied:
Would you like me to get rid of the color icons or maybe return a `map[string]string` with only the language and the color

silverwind replied:
Color icons are a nice touch, let's keep them.

lunny replied:
> We also need the Color, which we get with `LoadAttributes()`

If ids haven't been loaded from database, how does the `LoadAttributes` work?
And since there is `Distinct`, not all databases support loading ids. Except you add `id` as another `Distinct` columns.

JakobDev replied:
This is the current Code of `LoadAttributes()`:
```go
// LoadAttributes loads attributes
func (stats LanguageStatList) LoadAttributes() {
	for i := range stats {
		stats[i].Color = enry.GetColor(stats[i].Language)
	}
}
```
As you can see, it only uses the Language Column and not the ID Column. Using the ID Column for the Color would not make much sense, as this is different on every Database.

lunny commented on routers/web/explore/repo.go:
```diff
@@ -129,10 +129,17 @@ func RenderRepoSearch(ctx *context.Context, opts *RepoSearchOptions) {
 	ctx.Data["Total"] = count
 	ctx.Data["Repos"] = repos
 	ctx.Data["IsRepoIndexerEnabled"] = setting.Indexer.RepoIndexerEnabled
+	ctx.Data["Languages"] = languages
```
How about renaming this as `ProgramLanguages` to reduce the possibility of conflicting with UI Languages.

lunny commented:
I will send my LGTM now. But a further step is to use async menus with a search box because the programming languages may be very big on public websites.

JakobDev commented:
> But a further step is to use async menus with a search box because the programming languages may be very big on public websites

I think this should be done, before this is merged.

yp05327 commented on routers/web/explore/repo.go:
```diff
@@ -129,10 +129,17 @@ func RenderRepoSearch(ctx *context.Context, opts *RepoSearchOptions) {
 
+	languages, err := repo_model.GetPrimaryRepoLanguageList(ctx)
+	if err != nil {
+		ctx.ServerError("SearchRepository", err)
```
```suggestion
		ctx.ServerError("GetPrimaryRepoLanguageList", err)
```

snematoda commented:
> > But a further step is to use async menus with a search box because the programming languages may be very big on public websites
> 
> I think this should be done, before this is merged.

Added search for now

silverwind commented:
> I will send my LGTM now. But a further step is to use async menus with a search box because the programming languages may be very big on public websites.

Please define "big". The number after which to go async depends on the dropdown performance. I have other UI's where I don't consider pagination until 10k items. Test first how many items the dropdown can render performantly, and then decide whether pagination is worth it.

snematoda commented:
>  Test first how many items the dropdown can render performantly, and then decide whether pagination is worth it. 

Well, I don't think I have the resources to test on even a moderately sized instance.

However initially, rendering `.LanguagesColor` (~500 items) directly from enry (without db calls) seemed fine

silverwind commented:
Okay, if 500 entries render without performance issues, then we don't need async.

JakobDev commented:
You should probably switch the position of Sort and the Language Filter. Sorting is usually the last in Gitea.

snematoda commented:
> You should probably switch the position of Sort and the Language Filter. Sorting is usually the last in Gitea.

Done!

JakobDev commented on templates/explore/repo_search.tmpl:
```diff
@@ -1,4 +1,25 @@
+				<i class="icon gt-df gt-ac gt-jc">{{svg "octicon-search" 16}}</i>
+				<input type="text" placeholder="Language">
+			</div>
+			<a class="item" href="{{$.Link}}?q={{$.Keyword}}&sort={{$.SortType}}">Any</a>
```
You should set this to active, if no language is set

JakobDev reviewed (commented):
How long can the List get before a scrollbar appears?

snematoda commented:
> How long can the List get before a scrollbar appears?

500px, I believe that's the value for issue labels

snematoda commented:
> Should we maybe set this in relation to the size of the site, so it will not be to long?

I agree, just set the max-width to 60vh looks much better now

puni9869 commented:
> > Should we maybe set this in relation to the size of the site, so it will not be to long?
> 
> Just set the max-width to 60vh looks much better now

Could you provide a screenshot of that.

puni9869 commented on templates/explore/repo_search.tmpl:
```diff
@@ -1,4 +1,25 @@
-<div class="ui right floated secondary filter menu">
+<div class="ui right floated secondary filter menu" style="z-index: 1">
```
```suggestion
<div class="ui right floated secondary filter menu" ">
```
`style="z-index: 1` can you this adjust some how in css files in class

silverwind replied:
Your suggestion has unbalanced `"`. Also I agree we shoud ideally not need this z-index adjustment at all. What is the exact problem?

puni9869 commented on templates/explore/repo_search.tmpl:
```diff
@@ -1,4 +1,25 @@
+		<div class="menu" style="max-height: 60vh; overflow-y: auto;">
+			<div class="ui icon search input">
+				<i class="icon gt-df gt-ac gt-jc">{{svg "octicon-search" 16}}</i>
+				<input type="text" placeholder="Language">
```
```suggestion
				<input type="text" placeholder="Language">
```
`Language` may be you need to  translate it.

puni9869 commented on templates/explore/repo_search.tmpl:
```diff
@@ -21,7 +42,7 @@
 	</div>
 </div>
-<form class="ui form ignore-dirty" style="max-width: 90%">
+<form class="ui form ignore-dirty" style="max-width: 70%">
```
```suggestion
<form class="ui form ignore-dirty" >
```
`style="max-width: 70%"` please avoid the inline styles

silverwind replied:
We removed these already in master branch.

snematoda replied:
removed all the inline styles

snematoda commented:
> Your suggestion has unbalanced ". Also I agree we shoud ideally not need this z-index adjustment at all. What is the exact problem?

the search input would cover the dropdowns on smaller screens with 1.19, removed it as it seems to be fixed with 1.20

---

PS: Unsure if i messed it up, but I apologise for the force push

silverwind commented:
Small merge conflict in CSS.

snematoda commented:
> Small merge conflict in CSS

Should be fixed now

silverwind commented on templates/explore/repo_search.tmpl:
```diff
@@ -12,25 +12,48 @@
+		<!-- Language -->
+		<div class="ui right dropdown type jump item gt-mx-0">
+			<span class="text">
+				{{.locale.Tr "repo.repo_lang"}} 
```
```suggestion
				{{.locale.Tr "repo.repo_lang"}}
```

silverwind commented on web_src/css/explore.css:
```diff
@@ -10,6 +10,28 @@
+  .explore .filter > div {
+    align-self: end;
+    margin-top: .25em;
+  }
```
I see these `form` and `div` selectors likely to break during refactoring. Can you give them some unique classnames please?

silverwind commented on web_src/css/explore.css:
```diff
@@ -10,6 +10,28 @@
+
+.explore .filter .ui.dropdown {
+  float: right;
+}
```
Some unique classnames would be beneficial here as well. We should not tie CSS so tightly to HTML structure, because HTML structure can always change during refactoring, which is where these selectors break.

snematoda replied:
done :)

wxiaoguang commented on web_src/css/explore.css:
```diff
@@ -10,6 +10,29 @@
+
+.explore-lang,
+.explore-sort {
+  float: right;
```
If the page is using "flex" layout, why it needs "float: right" here?

silverwind replied:
I think previous sort element had `float` as well, but we can likely replace this with flex as well.

silverwind replied:
Will do a few CSS fixes here.

silverwind replied:
Flex done in https://github.com/go-gitea/gitea/pull/24212/commits/1e59ce2bb4fdc0620c98ed6aebda89cbbcb2e994, no more float.

snematoda replied:
> If the page is using "flex" layout, why it needs "float: right" here?

The parent element has display set to block.

I usually try to avoid nested flexboxes due to weird WebKit bugs which have been reported (for different projects) in the past.

would you like me change flex over float(will change if you prefer using flex instead)?

silverwind commented on templates/explore/repo_search.tmpl:
```diff
@@ -12,25 +12,48 @@
+				<a class="{{if not $.Language}}active {{end}}item" href="{{$.Link}}?q={{$.Keyword}}&sort={{$.SortType}}">Any</a>
+				{{range .ProgramLanguages}}
+				<a class="{{if eq $.Language .Language}}active {{end}}item" href="{{$.Link}}?q={{$.Keyword}}&language={{.Language}}&sort={{$.SortType}}">
+					<span class="gt-df gt-ac gt-mr-3"><i class="color-icon gt-mr-3" style="background-color: {{.Color}}"></i>{{.Language}}</span>
```
The class on the `span` is pointless, removing...

silverwind replied:
Fixed in https://github.com/go-gitea/gitea/pull/24212/commits/ab871331216be40989f009e8c5780c60cbcdfa1d:

<img width="298" alt="Screenshot 2023-08-14 at 15 53 04" src="https://github.com/go-gitea/gitea/assets/115237/8095d828-fe4f-44e2-b9c8-b2ff5a4b6031">

wxiaoguang commented:
There is one more problem.

Since this template is shared by different pages, now, on the "User Profile" page, the list is always empty.


![image](https://github.com/go-gitea/gitea/assets/2114189/309ae0ca-f50f-48c7-bbbd-1c4393c4cde9)

JakobDev commented on templates/explore/repo_search.tmpl:
```diff
@@ -12,25 +12,48 @@
+						<input type="text" placeholder="{{.locale.Tr "repo.repo_lang"}}">
+					</div>
+					<a class="{{if not $.Language}}active {{end}}item" href="{{$.Link}}?q={{$.Keyword}}&sort={{$.SortType}}">
+						<span class="gt-df gt-ac gt-gap-3"><i class="color-icon gt-border-secondary" style="background-color: inherit;"></i>Any</span>
```
Any should be translatable

snematoda replied:
Switched to the already translated `any`

snematoda commented:
> There is one more problem.

> Since this template is shared by different pages, now, on the "User Profile" page, the list is always empty.

@wxiaoguang sorry about that, I'm quite new to gitea's codebase ... I have added support for user and org too now, for other cases it's currently hidden

Note: currently in user/org all the languages used in the instance are displayed (since the `language_stat` table does not contain an `owner_id` field, unsure if searching in the repos is scalable)

wxiaoguang commented:
Never mind (no need to "sorry" 😄 ) , Gitea's code base is quite complex .....

But for the "language" filter on the user's profile page .... is it better to just hide it at the moment? The global language list doesn't seem fit the user's repo list.

snematoda commented:
>  But for the "language" filter on the user's profile page .... is it better to just hide it at the moment? The global language list doesn't seem to fit the user's repo list.

True :)

I'll try creating the langs from the already fetched repos first, it wouldn't list all the languages used by the user but it's better than displaying nothing :P

silverwind commented on routers/api/v1/repo/languages.go:
```diff
@@ -0,0 +1,54 @@
+	}
+
+	ctx.JSON(http.StatusOK, list)
+}
```
```suggestion
}

```

silverwind commented:
@JakobDev looking good?

JakobDev commented:
The API needs a little bit of improvement. If you want to get all languages for a User you need o load all Repos of that User and pass the IDs to `/repos/languages`. I think we should also add `/user/{username}/languages` and `/org/{orgs}/languages` to get the languages for a User/Org.

We need also tests for the API to make sure it keeps working as expected. If you don't know how to write tests, I can do that for you.

JakobDev commented on routers/api/v1/repo/languages.go:
```diff
@@ -0,0 +1,54 @@
+		repos = nil
+	}
+
+	langs, err := repo_model.GetPrimaryRepoLanguageList(db.DefaultContext, repos)
```
```suggestion
	langs, err := repo_model.GetPrimaryRepoLanguageList(ctx, repos)
```

lafriks commented on models/repo/repo_list.go:
```diff
@@ -735,3 +735,31 @@ func GetUserRepositories(opts *SearchRepoOptions) (RepositoryList, int64, error)
+func GetPrimaryRepoLanguageList(ctx context.Context, repos RepositoryList) (LanguageStatList, error) {
+	languageList := make(LanguageStatList, 0)
+
+	q := db.GetEngine(ctx).
```
I don't think this will perform well in large instances. It would be better to get this data form FTS

snematoda replied:
Discussed at https://github.com/go-gitea/gitea/pull/24212#discussion_r1172249016

Unsure what the ideal alternative would be

lunny replied:
Maybe add a new table to store the languages and counts

snematoda replied:
> Maybe add a new table to store the languages and counts

Hmm, might work for the global list but unsure how to  filter them by the repo owner since a few databases do not support storing primitive types as an array. Might be possible to create yet another table containing a the list of owners but tbf I wonder if the benefits would be worth it at that point

snematoda commented:
> The API needs a little bit of improvement. If you want to get all languages for a User you need o load all Repos of that User and pass the IDs to /repos/languages. I think we should also add /user/{username}/languages and /org/{orgs}/languages to get the languages for a User/Org.

@JakobDev I have added a query param for the ownerId (ignored when repo ids are provided fn) ... felt it was easier than adding multiple api routes :P

> We need also tests for the API to make sure it keeps working as expected. If you don't know how to write tests, I can do that for you.

Thank you! I have added in a minimal test fn

puni9869 commented on web_src/fomantic/build/semantic.css:
```diff
@@ -3689,6 +3689,11 @@ body.dimmable > .dimmer {
   will-change: transform, opacity;
 }
 
+.ui.dropdown .menu.short {
+  max-width: 60vh;
+  overflow-y: auto;
+}
+
```
this can be moved to base.css

snematoda replied:
Fixed in https://github.com/go-gitea/gitea/pull/24212/commits/00fb639246c795673bb1bdaa95858827316504be

puni9869 commented on modules/context/base.go:
```diff
@@ -198,20 +198,6 @@ func (b *Base) FormInt64(key string) int64 {
-	}
-	return v
-}
-
```
This is a breaking change. `FormInt64s` may be shared among many places.

puni9869 replied:
<img width="694" alt="image" src="https://github.com/go-gitea/gitea/assets/80308335/e2636047-fafe-48bc-ba1b-5b2bf9c8aae1">

snematoda replied:
`FormInt64` still exists, `FormInt64s` was added in https://github.com/go-gitea/gitea/pull/24212/commits/e50b52ccd08ab7cbe7e7691d7c86da5a59551864#diff-fa4a8538493d18cf504db405186f872622f67001de4bd2ecb5807d7e5408ccfd as a int64 counterpart for `FormStrings` in this PR and was only used for fetching this ids in the api. But it has since been removed since the ids was replaced by proper user/org support in the last commit :)

JakobDev commented on models/repo/repo_list.go:
```diff
@@ -745,3 +745,34 @@ func GetUserRepositories(opts *SearchRepoOptions) (RepositoryList, int64, error)
+	if ownerID > 0 {
+		ids, err := SearchRepositoryIDsByCondition(ctx, builder.NewCond().And(
+			builder.Eq{"owner_id": ownerID},
+			builder.Eq{"is_private": private},
```
This will not work as you expect. You can give other users access to a private repo. if you do this, they will see the private repo. So you need to check for each repo if the current user has read access. Take a look at  `repo_model.SearchRepositoryByName` and `repo_model.CountRepository` which both o the check.

snematoda replied:
Eek, I should have tested out with multiple accounts

JakobDev commented on tests/integration/api_languages_test.go:
```diff
@@ -0,0 +1,47 @@
+	DecodeJSON(t, resp, &langs)
+
+	for _, lang := range langs {
+		assert.NotEqual(t, lang.Name, "")
+		assert.NotEqual(t, lang.Color, "")
```
You should use Equal and not check, if the String is not Empty.

lunny commented:
@snematoda Can you send the PR again? I want to resolve the conflicts but made the PR broken.
